### ISSUE TITLE
Update of exiting OE fails, adding a new OE fails

### ISSUE THREAD

**Issue by smuellerDD** [usnistgov/ACVP-Server#65]
**environment**
Demo

**testSessionId**
N/A

**vsId**
N/A

The ACVP server DB contains an OE ID: 31749. Its contents contain:

```
{
  "url":"/acvp/v1/oes/31749",
  "name":"Fedora 29 on Intel Broadwell i7 on MyOwn MyOwnSeries MyOwnName",
  "dependencies":[
    {
      "url":"/acvp/v1/dependencies/20094",
      "type":"processor",
      "name":"i7",
      "description":"Processor i7 (processor family X86) from Intel",
      "manufacturer":"Intel",
      "family":"X86",
      "series":"Broadwell"
    },
    {
      "url":"/acvp/v1/dependencies/22922",
      "type":"software",
      "name":"Fedora 29",
      "description":"Linux distribution developed by the community-supported Fedora Project and sponsored by Red Hat, Inc.",
      "swid":"org.fedoraproject.Fedora-29"
    },
    {
      "url":"/acvp/v1/dependencies/23630",
      "type":"processor",
      "name":"MyOwnName",
      "description":"Processor MyOwnName (processor family X86) from MyOwn",
      "manufacturer":"MyOwn",
      "family":"X86",
      "series":"MyOwnSeries"
    },
    {
      "url":"/acvp/v1/dependencies/23640",
      "type":"software",
      "name":"Fedora 29",
      "description":"Linux distribution developed by the community-supported Fedora Project and sponsored by Red Hat, Inc.",
      "swid":"org.fedoraproject.Fedora-29"
    },
    {
      "url":"/acvp/v1/dependencies/23641",
      "type":"processor",
      "name":"i7",
      "description":"Processor i7 (processor family X86) from Intel",
      "manufacturer":"Intel",
      "family":"X86",
      "series":"Broadwell"
    },
    {
      "url":"/acvp/v1/dependencies/23642",
      "type":"processor",
      "name":"MyOwnName",
      "description":"Processor MyOwnName (processor family X86) from MyOwn",
      "manufacturer":"MyOwn",
      "family":"X86",
      "series":"MyOwnSeries"
    }
  ]
}
```

I would like to change it and send a PUT with the following payload:

```
{
  "name":"Fedora 29 on Intel Broadwell i7 on MyOwn MyOwnSeries MyOwnName",
  "dependencies":[
    {
      "type":"software",
      "name":"Fedora 29",
      "cpe":null,
      "swid":"org.fedoraproject.Fedora-29",
      "description":"Linux distribution developed by the community-supported Fedora Project and sponsored by Red Hat, Inc."
    },
    {
      "type":"processor",
      "manufacturer":"Intel",
      "family":"X86",
      "name":"i7",
      "series":"Broadwell",
      "description":"Processor i7 (processor family X86) from Intel"
    },
    {
      "type":"processor",
      "manufacturer":"MyOwn",
      "family":"X86",
      "name":"MyOwnName",
      "series":"MyOwnSeries",
      "description":"Processor MyOwnName (processor family X86) from MyOwn"
    }
  ],
  "dependencyUrls":[
    "/acvp/v1/dependencies/22922",
    "/acvp/v1/dependencies/20094",
    "/acvp/v1/dependencies/23630"
  ]
}
```

The server responds to me with giving a request ID: 17656

Eventually it will be approved:

```
  {
    "url": "/acvp/v1/requests/17656",
    "status": "approved",
    "approvedUrl": "/acvp/v1/oes/31749"
  }
```

After the approval, I try to check the DB contents for the OE ID 31749 and I get the old content:

```
[
  {
    "acvVersion": "1.0"
  },
  {
    "url": "/acvp/v1/oes/31749",
    "name": "Fedora 29 on Intel Broadwell i7 on MyOwn MyOwnSeries MyOwnName",
    "dependencies": [
      {
        "url": "/acvp/v1/dependencies/20094",
        "type": "processor",
        "name": "i7",
        "description": "Processor i7 (processor family X86) from Intel",
        "manufacturer": "Intel",
        "family": "X86",
        "series": "Broadwell"
      },
      {
        "url": "/acvp/v1/dependencies/22922",
        "type": "software",
        "name": "Fedora 29",
        "description": "Linux distribution developed by the community-supported Fedora Project and sponsored by Red Hat, Inc.",
        "swid": "org.fedoraproject.Fedora-29"
      },
      {
        "url": "/acvp/v1/dependencies/23630",
        "type": "processor",
        "name": "MyOwnName",
        "description": "Processor MyOwnName (processor family X86) from MyOwn",
        "manufacturer": "MyOwn",
        "family": "X86",
        "series": "MyOwnSeries"
      },
      {
        "url": "/acvp/v1/dependencies/23643",
        "type": "software",
        "name": "Fedora 29",
        "description": "Linux distribution developed by the community-supported Fedora Project and sponsored by Red Hat, Inc.",
        "swid": "org.fedoraproject.Fedora-29"
      },
      {
        "url": "/acvp/v1/dependencies/23644",
        "type": "processor",
        "name": "i7",
        "description": "Processor i7 (processor family X86) from Intel",
        "manufacturer": "Intel",
        "family": "X86",
        "series": "Broadwell"
      },
      {
        "url": "/acvp/v1/dependencies/23645",
        "type": "processor",
        "name": "MyOwnName",
        "description": "Processor MyOwnName (processor family X86) from MyOwn",
        "manufacturer": "MyOwn",
        "family": "X86",
        "series": "MyOwnSeries"
      }
    ]
  }
]
```

This happens consistently: I cannot update the DB.

Now, I try to register the OE as a *new* entry by sending a POST message to ../oes. Once the server approves the request, I get a new OE ID: 

```
  {
    "url": "/acvp/v1/requests/17657",
    "status": "approved",
    "approvedUrl": "/acvp/v1/oes/31750"
  }
```
I.e. the new OE ID is *different* than above.

Yet, after validating the newly registered OE ID, I get the same old content for the *new* ID:

```
[
  {
    "acvVersion": "1.0"
  },
  {
    "url": "/acvp/v1/oes/31750",
    "name": "Fedora 29 on Intel Broadwell i7 on MyOwn MyOwnSeries MyOwnName",
    "dependencies": [
      {
        "url": "/acvp/v1/dependencies/20094",
        "type": "processor",
        "name": "i7",
        "description": "Processor i7 (processor family X86) from Intel",
        "manufacturer": "Intel",
        "family": "X86",
        "series": "Broadwell"
      },
      {
        "url": "/acvp/v1/dependencies/22922",
        "type": "software",
        "name": "Fedora 29",
        "description": "Linux distribution developed by the community-supported Fedora Project and sponsored by Red Hat, Inc.",
        "swid": "org.fedoraproject.Fedora-29"
      },
      {
        "url": "/acvp/v1/dependencies/23630",
        "type": "processor",
        "name": "MyOwnName",
        "description": "Processor MyOwnName (processor family X86) from MyOwn",
        "manufacturer": "MyOwn",
        "family": "X86",
        "series": "MyOwnSeries"
      },
      {
        "url": "/acvp/v1/dependencies/23646",
        "type": "software",
        "name": "Fedora 29",
        "description": "Linux distribution developed by the community-supported Fedora Project and sponsored by Red Hat, Inc.",
        "swid": "org.fedoraproject.Fedora-29"
      },
      {
        "url": "/acvp/v1/dependencies/23647",
        "type": "processor",
        "name": "i7",
        "description": "Processor i7 (processor family X86) from Intel",
        "manufacturer": "Intel",
        "family": "X86",
        "series": "Broadwell"
      },
      {
        "url": "/acvp/v1/dependencies/23648",
        "type": "processor",
        "name": "MyOwnName",
        "description": "Processor MyOwnName (processor family X86) from MyOwn",
        "manufacturer": "MyOwn",
        "family": "X86",
        "series": "MyOwnSeries"
      }
    ]
  }
]
```

Bottom line: I cannot create/update my OE.

**Comment by Kritner** [comment]
@smuellerDD I've taken a look at `oes/31749` and I'm not seeing what you're stating as the issue.  

From your issue above, after the "initial create" of 31749, you had dependency IDs of:

```
20094
22922
23630
23640
23641
23642
```

On your update, you state you're attempting to create 3 new dependencies, and reuse the following dependencies:

```
22922
20094
23630
```

the above contains 3 IDs from the "initial posting" of the creation of the OE, with 3 "new dependencies" - which as far as I can tell, are describing the same 3 dependencies, just as "new ones" rather than reusing the ids 23640, 23641, and 23642 from the initial post.

Your eventual response to the `PUT` operation gives back the following IDs:

```
20094
22922
23630
23643
23644
23645
```

The top 3 dependencies are the IDs you specified to reuse in your put operation, and the last 3 are the new "in line" dependencies that you provided.

I'm really not sure where you're saying the error is occurring here, can you elaborate?  Also I'm not sure where to start regarding the secondary post message, since you didn't specify the contents of that post.  If you simply changed the "PUT" payload to a "POST", then you're *also* getting back the response that you should be getting back, as far as I can tell.

I'm not sure what exactly you're trying to accomplish with these POSTs and PUTs, and I think getting that clarification will help me dig into the issue.

**Comment by smuellerDD** [comment]
Just tried again.

On server:

```
{
  "url":"/acvp/v1/oes/31750",
  "name":"Fedora 29 on Intel Broadwell i7 on MyOwn MyOwnSeries MyOwnName",
  "dependencies":[
    {
      "url":"/acvp/v1/dependencies/20094",
      "type":"processor",
      "name":"i7",
      "description":"Processor i7 (processor family X86) from Intel",
      "manufacturer":"Intel",
      "family":"X86",
      "series":"Broadwell"
    },
    {
      "url":"/acvp/v1/dependencies/22922",
      "type":"software",
      "name":"Fedora 29",
      "description":"Linux distribution developed by the community-supported Fedora Project and sponsored by Red Hat, Inc.",
      "swid":"org.fedoraproject.Fedora-29"
    },
    {
      "url":"/acvp/v1/dependencies/23630",
      "type":"processor",
      "name":"MyOwnName",
      "description":"Processor MyOwnName (processor family X86) from MyOwn",
      "manufacturer":"MyOwn",
      "family":"X86",
      "series":"MyOwnSeries"
    },
    {
      "url":"/acvp/v1/dependencies/23655",
      "type":"software",
      "name":"Fedora 29",
      "description":"Linux distribution developed by the community-supported Fedora Project and sponsored by Red Hat, Inc.",
      "swid":"org.fedoraproject.Fedora-29"
    },
    {
      "url":"/acvp/v1/dependencies/23656",
      "type":"processor",
      "name":"i7",
      "description":"Processor i7 (processor family X86) from Intel",
      "manufacturer":"Intel",
      "family":"X86",
      "series":"Broadwell"
    },
    {
      "url":"/acvp/v1/dependencies/23657",
      "type":"processor",
      "name":"MyOwnName",
      "description":"Processor MyOwnName (processor family X86) from MyOwn",
      "manufacturer":"MyOwn",
      "family":"X86",
      "series":"MyOwnSeries"
    }
  ]
}
```

My PUT operation:

```
[{"acvVersion":"1.0"},{"name":"Fedora 29 on Intel Broadwell i7 on MyOwn MyOwnSeries MyOwnName","dependencyUrls":["/acvp/v1/dependencies/22922","/acvp/v1/dependencies/20094","/acvp/v1/dependencies/23630"]}]
```

And now it works.
